README.md
# woocommerceaio

An async Python wrapper for the WooCommerce REST API based on [httpx](https://www.python-httpx.org/). 

This is an _unofficial_ fork of the [WooCommerce-provided Python client](https://github.com/woocommerce/wc-api-python), originally created by Claudio Sanches.

The main difference between `woocommerceaio` and the official client resides in the use of the async HTTP library [httpx](https://www.python-httpx.org/) instead of [requests](https://requests.readthedocs.io/en/latest/).

Other differences include:

- Support for retries.
- Type hints.
- And more to come (as I use the library I need to add more features)

## Installation

```sh
pip install woocommerceaio
```

## Getting started

Generate API credentials (Consumer Key & Consumer Secret) following this instructions at http://woocommerce.github.io/woocommerce-rest-api-docs/#rest-api-keys.

Check out the WooCommerce API endpoints and data that can be manipulated in http://woocommerce.github.io/woocommerce-rest-api-docs/.

## Setup

```python
    from woocommerceaio import API

    wcapi = API(
        url="http://example.com",
        consumer_key="ck_XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
        consumer_secret="cs_XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
        version="wc/v3"
    )
```

## Options

| Option | Type | Required | Description |
|--------|------|----------|-------------|
| `url` | `string` | yes | Your Store URL, example: http://woo.dev/ |
| `consumer_key` | `string`  | yes | Your API secret key |
| `consumer_secret` | `string` | yes | Your API consumer secret |
| `version` | `string` | no | API version, default is `wc/v3` |
| `timeout` | `integer` | no | Connection timeout, default is `5` |
| `verify_ssl` | `bool` | no | Verify SSL when connect, use this option as `False` when need to test with self-signed certificates |
| `query_string_auth` | `bool` | no | Force Basic Authentication as query string when `True` and using under HTTPS, default is `False` |
| `user_agent` | `string` | no | Set a custom User-Agent, default is `woocommerceaio/<version>` |
| `oauth_timestamp` | `integer` | no | Custom timestamp for requests made with oAuth1.0a |
| `wp_api` | `bool` | no | Set to `False` in order to use the legacy WooCommerce API (deprecated) |

## Methods

You can interact with the WooCommerce API using via the exposed methods `get()`, `post()`, `put()`, `delete()`, `options()`.

- All methods

 the following parameters.

| Params | Type | Description |
|--------|------|-------------|
| `endpoint` | `string` | WooCommerce API endpoint, example: `customers` or `order/12` |
| `data` | `dictionary` | Data that will be converted to JSON |
| `**kwargs` | `dictionary` | Accepts `params`, also other `httpx` arguments (see next section) |

### kwargs params

`woocommerceaio` allows you to pass any `httpx` arguments as `kwargs` params. This is useful if you want to customize headers, timeouts, and other request parameters. For a list of available params, refer to the [httpx documentation](https://www.python-httpx.org/api/#request).

Additionally, `woocommerceaio` provides the following custom `kwargs` params:

| Params | Type | Description |
|--------|------|-------------|
| `max_retries` | `bool` | Maximum number of retries on failed requests. Default is `3` |
| `retry_backoff` | `float` | Backoff time to apply between attempts after the second try (most errors are resolved immediately by a second try without a delay). Default is `0.5` seconds. This value will be multiplied by a factor of two for each consequent retry. So the default delay sequence will be `0.5`, `1`, `2`, `4`, `8` etc. seconds |

### Response

All methods will return an httpx [`Response`](https://www.python-httpx.org/api/#response) object. For more information on how to use this object, refer to the [httpx documentation](https://www.python-httpx.org/api/#response).

## Examples

### Retrieving store products

```python
    >>> r = await wcapi.get("products")
    >>> r.status_code
    200
    >>> r.headers['content-type']
    'application/json; charset=UTF-8'
    >>> r.encoding
    'UTF-8'
    >>> r.text
    u'{"products":[{"title":"Flying Ninja","id":70,...' // Json text
    >>> r.json()
    {u'products': [{u'sold_individually': False,... // Dictionary data
```

### Making requests with `params`

```python
    from woocommerceaio import API

    async def example():
        wcapi = API(
            url="http://example.com",
            consumer_key="ck_XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
            consumer_secret="cs_XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX",
            version="wc/v3"
        )

        # Force delete example.
        response = await wcapi.delete("products/100", params={"force": True})
        print(response.json())

        # Query example.
        response = await wcapi.get("products", params={"per_page": 20})
        print(response.json())

    if __name__ == "__main__":
        import asyncio
        loop = asyncio.get_event_loop()
        loop.run_until_complete(example())
```

## Changelog

See `CHANGELOG.md`.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.1.3] - 2023-11-19

### Changed

- Allow to specify a backoff time for retries with `retry_backoff`.
- Improved the documentation in the README file.

## [1.1.2] - 2023-11-18

### Fixed

- Improper handling of status code 429 (Too Many Requests).

## [1.1.1] - 2023-11-18

### Changed

- Raise `httpx.TimeoutException` instead of a generic `Exception` on max retries exceeded.

## [1.1.0] - 2023-11-17

### Added

- Support for retries on failed requests (param `max_retries`)

## [1.0.1] - 2023-11-12

### Added

- `py.typed` marker file.

## [1.0.0] - 2023-11-12

### Added

- async API based on the original code in [https://github.com/woocommerce/wc-api-python](https://github.com/woocommerce/wc-api-python).


src/woocommerceaio/api.py
__version__ = "1.1.3"

import asyncio
import logging
from httpx import AsyncClient, Response, BasicAuth, HTTPError, TimeoutException
from json import dumps as jsonencode
from time import time
from woocommerceaio.oauth import OAuth
from urllib.parse import urlencode

import typing as t


class API(object):
    """API Class"""

    def __init__(
        self,
        url: str,
        consumer_key: str,
        consumer_secret: str,
        max_retries: int = 3,
        timeout: int = 5,
        **kwargs,
    ):
        self.url = url
        self.consumer_key = consumer_key
        self.consumer_secret = consumer_secret
        self.wp_api = kwargs.get("wp_api", True)
        self.version = kwargs.get("version", "wc/v3")
        self.is_ssl = self.__is_ssl()
        self.max_retries = max_retries
        self.timeout = timeout
        self.verify_ssl = kwargs.get("verify_ssl", True)
        self.query_string_auth = kwargs.get("query_string_auth", False)
        self.user_agent = kwargs.get("user_agent", f"woocommerceaio/{__version__}")

    def __is_ssl(self) -> bool:
        """Check if url use HTTPS"""
        return self.url.startswith("https")

    def __get_url(self, endpoint: str) -> str:
        """Get URL for requests"""
        url = self.url
        api = "wc-api"

        if url.endswith("/") is False:
            url = f"{url}/"

        if self.wp_api:
            api = "wp-json"

        return f"{url}{api}/{self.version}/{endpoint}"

    def __get_oauth_url(self, url: str, method: str, **kwargs: t.Any) -> str:
        """Generate oAuth1.0a URL"""
        oauth = OAuth(
            url=url,
            consumer_key=self.consumer_key,
            consumer_secret=self.consumer_secret,
            version=self.version,
            method=method,
            oauth_timestamp=kwargs.get("oauth_timestamp", int(time())),
        )

        return oauth.get_oauth_url()

    async def __request(
        self,
        method: str,
        endpoint: str,
        data: t.Optional[t.Any],
        params: t.Optional[t.Dict[str, t.Any]] = None,
        **kwargs: t.Any,
    ) -> Response:
        """Do requests"""
        if params is None:
            params = {}
        url = self.__get_url(endpoint)
        auth = None
        headers = {"user-agent": f"{self.user_agent}", "accept": "application/json"}

        if self.is_ssl is True and self.query_string_auth is False:
            auth = BasicAuth(self.consumer_key, self.consumer_secret)
        elif self.is_ssl is True and self.query_string_auth is True:
            params.update(
                {
                    "consumer_key": self.consumer_key,
                    "consumer_secret": self.consumer_secret,
                }
            )
        else:
            encoded_params = urlencode(params)
            url = f"{url}?{encoded_params}"
            url = self.__get_oauth_url(url, method, **kwargs)

        if data is not None:
            data = jsonencode(data, ensure_ascii=False).encode("utf-8")
            headers["content-type"] = "application/json;charset=utf-8"

        follow_redirects: bool = False
        if "allow_redirects" in kwargs:
            follow_redirects = kwargs.pop("allow_redirects")

        async with AsyncClient(
            verify=self.verify_ssl,
            follow_redirects=follow_redirects,
            timeout=self.timeout,
            headers=headers,
            auth=auth,
        ) as client:
            backoff: float = (
                float(kwargs.pop("retry_backoff")) if "retry_backoff" in kwargs else 0.5
            )
            _try: int = 1

            while True:
                try:
                    response: Response = await client.request(
                        method=method,
                        url=url,
                        params=params,
                        content=data,
                        **kwargs,
                    )
                    if response.status_code < 500 and response.status_code != 429:
                        return response
                except HTTPError as ex:
                    logging.error(
                        f"Error retrieving {self.__get_url(endpoint)}: {str((ex))}"
                    )

                if _try == self.max_retries:
                    break
                logging.error(f"Awaiting {backoff} seconds before next attempt...")
                await asyncio.sleep(backoff)
                _try += 1
                backoff *= 2

            raise TimeoutException(
                f"Max retries exceeded to {self.__get_url(endpoint)}"
            )

    async def get(self, endpoint: str, **kwargs: t.Any) -> Response:
        """Get requests"""
        return await self.__request("GET", endpoint, None, **kwargs)

    async def post(
        self, endpoint: str, data: t.Optional[t.Dict[str, t.Any]], **kwargs: t.Any
    ) -> Response:
        """POST requests"""
        return await self.__request("POST", endpoint, data, **kwargs)

    async def put(
        self, endpoint: str, data: t.Optional[t.Dict[str, t.Any]], **kwargs: t.Any
    ) -> Response:
        """PUT requests"""
        return await self.__request("PUT", endpoint, data, **kwargs)

    async def delete(self, endpoint: str, **kwargs) -> Response:
        """DELETE requests"""
        return await self.__request("DELETE", endpoint, None, **kwargs)

    async def options(self, endpoint: str, **kwargs) -> Response:
        """OPTIONS requests"""
        return await self.__request("OPTIONS", endpoint, None, **kwargs)
